codecov_auth/views/okta_mixin.py

```diff
@@ -0,0 +1,112 @@
+import json
+import logging
+import re
+from urllib.parse import urlencode
+
+import jwt
+import pydantic
+import requests
+from django.http import HttpResponse
+from django.shortcuts import redirect
+from requests.auth import HTTPBasicAuth
+
+from codecov_auth.views.base import StateMixin
+
+log = logging.getLogger(__name__)
+
+ISS_REGEX = re.compile(r"https://[\w\d\-\_]+.okta.com/?")
+
+
+class OktaTokenResponse(pydantic.BaseModel):
```

Comment:
What are your thoughts on classes vs dataclasses? I tend to go back and forth w/ dataclasses; my goal at the end of the day is to know what an object's shape is going to look like, and this does this which is totally fine. Curious to hear if you have a strong opinion for/against the way you type things

Reply:
I’m also a bit surprised to see it like this, as it is pulling in `pydantic` as external dependency. Does this offer some more safety features vs just using plain `mypy`?

Reply:
Pydantic also does runtime type checking (i.e. ensures that 1. the fields exist in the response and that 2. that they're string typed, or can be strings). `dataclass.dataclasses` doesn't do type checking (it just defines the structure) and `mypy` only checks for types when you run `mypy` (or "compile" time type checking... pydantic is more similar to rust's [serde lib](https://serde.rs/)). `pydantic` _might_ be overkill for this, but I ended up using this because `dataclass.dataclass` doesn't let me ignore fields that I don't care about (and I didn't want to recreate the whole datastructure of the response). In terms of why I used `BaseModel` instead of `pydantic.dataclasses`... I don't really have a good reason other than it's probably a bit more flexible? Pydantic `BaseModels` offer JSON dumping, which is also kind of useful if we want to convert back to `dicts` (i.e. sessions need to store serialized objects, for instance). [Here](https://docs.pydantic.dev/latest/concepts/dataclasses/) is a doc on some of the differences.